src/language-services/validation.ts

```diff
@@ -127,8 +180,151 @@ function maybeParseErrorToDiagnostic(error: PQP.ParseError.ParseError): Diagnost
     }
 
     return {
+        code: DiagnosticErrorCode.ParserError,
         message,
-        severity: DiagnosticSeverity.Error,
         range,
+        severity: DiagnosticSeverity.Error,
+        source,
     };
 }
+
+interface TraversalState extends PQP.Traverse.IState<Diagnostic[]> {
+    readonly documentUri: string;
+    readonly nodeIdMapCollection: PQP.NodeIdMap.Collection;
+    readonly source?: string;
+}
+
+function tryTraverse(
+    documentUri: string,
+    root: PQP.TXorNode,
+    nodeIdMapCollection: PQP.NodeIdMap.Collection,
+    options?: ValidationOptions,
+): PQP.Traverse.TriedTraverse<Diagnostic[]> {
+    const locale: string = LanguageServiceUtils.getLocale(options);
+    const localizationTemplates: PQP.ILocalizationTemplates = PQP.getLocalizationTemplates(locale);
+
+    const traversalState: TraversalState = {
+        documentUri,
+        localizationTemplates,
+        nodeIdMapCollection,
+        result: [],
+        source: options?.source,
+    };
+
+    return PQP.Traverse.tryTraverseXor(
+        traversalState,
+        nodeIdMapCollection,
+        root,
+        PQP.Traverse.VisitNodeStrategy.BreadthFirst,
+        visitNode,
+        PQP.Traverse.expectExpandAllXorChildren,
+        undefined,
+    );
+}
+
+function keyValuePairsToSymbols(
+    keyValuePairs: ReadonlyArray<
+        PQP.NodeIdMapIterator.KeyValuePair<PQP.Language.Ast.GeneralizedIdentifier | PQP.Language.Ast.Identifier>
+    >,
+): DocumentSymbol[] {
+    const symbols: DocumentSymbol[] = [];
+    keyValuePairs.forEach(value => {
```

Comment:
nit: use `for (const x of y)` instead of `foreach`. In most cases there's a small performance improvement.